synapse/handlers/saml_handler.py

```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
+    # But in case of crash we may have multiple sessions stored in the
+    # local cache. In that case we select the last session and remove
+    # other ones.
+    def _find_session_from_user(self, username):
+        subjects = self._saml_client.users.subjects()
+        sessions = []
+        # Find all sessions for a specific subject
+        for name_id in subjects:
+            # We should always have one source only
+            for source in self._saml_client.users.sources(name_id):
+                info = self._saml_client.users.get_info_from(name_id, source)
+                # If the username is found, append the session info
+                if (
+                    "ava" in info
+                    and "uid" in info["ava"]
+                    and username in info["ava"]["uid"]
+                ):
+                    info["entity"] = source
+                    sessions.append(info)
+
+        sessions.sort(key=lambda i: i["not_on_or_after"])
+        try:
+            # Retrieve last session
+            last_session = sessions.pop()
+            # We should have only one session active for one person
+            # This should never match but removing these staled sessions anyway
+            [self._saml_client.local_logout(session["name_id"]) for session in sessions]
+            return last_session
+        except IndexError:
+            return None
+
+    def _find_mxid_from_name_id(self, name_id):
+        try:
+            attributes = self._saml_client.users.get_identity(name_id)
+            for attribute in attributes:
+                if self._mxid_source_attribute in attribute:
+                    return "@%s:%s" % (
+                        attribute[self._mxid_source_attribute][0],
+                        self._auth_handler.hs.hostname,
+                    )
+        except Exception:
+            pass
+        return None
+
+    async def _logout(self, mxid):
+        # first delete all of the user's devices
+        await self._device_handler.delete_all_devices_for_user(mxid)
+
+        # .. and then delete any access tokens which weren't associated with
+        # devices.
+        await self._auth_handler.delete_access_tokens_for_user(mxid)
+
+    # Example: https://github.com/IdentityPython/pysaml2/blob/master/example/sp-wsgi/sp.py
+    def create_logout_request(self, user, access_token):
+        """Create a SAML logout request using HTTP redirect binding
+
+        Returns:
+            bytes: URL to redirect to
+        """
+        logger.info("Creating SAML logout request for %s", user)
+        try:
+            localpart = UserID.from_string(user).localpart
+            logger.debug("User localpart is %s", localpart)
+
+            session = self._find_session_from_user(localpart)
+            # The user probally logged in via m.login.password
+            if session is None:
+                return False
+            logger.debug("User session is %s", session)
```

Comment:
in your `debug` lines, can you try to give more context as to what it is you are showing, so that the admin has a better chance of understanding them when they are resolving problems?

for example: `"Found SAML2 session %s"`

Reply:
Corrected